File: assignments/objective-c/anagram/AnagramTest.m
#import <XCTest/XCTest.h>
#import "Anagram.h"

@interface test_suite : XCTestCase

@end

@implementation test_suite

- (void)testNoMatches {
  Anagram *anagram = [[Anagram alloc] initWithString:@"diaper"];
  NSArray *results = [anagram match:@[@"hello",@"world",@"zombies",@"pants"]];
  NSArray *expected = @[];

  XCTAssert([results isEqualToArray:expected]);
}

- (void)testDetectSimpleAnagram {
  Anagram *anagram = [[Anagram alloc] initWithString:@"ant"];
  NSArray *results = [anagram match:@[@"tan",@"stand",@"at"]];
  NSArray *expected = @[@"tan"];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testDetectMultipleAnagrams {
  Anagram *anagram = [[Anagram alloc] initWithString:@"master"];
  NSArray *results = [anagram match:@[@"stream",@"pigeon",@"maters"]];
  NSArray *expected = @[@"stream",@"maters"];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testDoesNotConfuseDifferentDuplicates {
  Anagram *anagram = [[Anagram alloc] initWithString:@"galea"];
  NSArray *results = [anagram match:@[@"eagle"]];
  NSArray *expected = @[];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testIdenticalWordIsNotAnagram {
  Anagram *anagram = [[Anagram alloc] initWithString:@"corn"];
  NSArray *results = [anagram match:@[@"corn", @"dark", @"Corn", @"rank", @"CORN", @"cron", @"park"]];
  NSArray *expected = @[@"cron"];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testEliminateAnagramsWithSameChecksum {
  Anagram *anagram = [[Anagram alloc] initWithString:@"mass"];
  NSArray *results = [anagram match:@[@"last"]];
  NSArray *expected = @[];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testEliminateAnagramSubsets {
  Anagram *anagram = [[Anagram alloc] initWithString:@"good"];
  NSArray *results = [anagram match:@[@"dog",@"goody"]];
  NSArray *expected = @[];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testDetectAnagram {
  Anagram *anagram = [[Anagram alloc] initWithString:@"listen"];
  NSArray *results = [anagram match:@[@"enlists",@"google",@"inlets",@"banana"]];
  NSArray *expected = @[@"inlets"];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testMultipleAnagrams {
  Anagram *anagram = [[Anagram alloc] initWithString:@"allergy"];
  NSArray *results = [anagram match:@[@"gallery",@"ballerina",@"regally",@"clergy",@"largely",@"leading"]];
  NSArray *expected = @[@"gallery",@"regally",@"largely"];
  XCTAssert([results isEqualToArray:expected]);
}

- (void)testAnagramsAreCaseInsensitive {
  Anagram *anagram = [[Anagram alloc] initWithString:@"Orchestra"];
  NSArray *results = [anagram match:@[@"cashregister",@"Carthorse",@"radishes"]];
  NSArray *expected = @[@"Carthorse"];
  XCTAssert([results isEqualToArray:expected]);
}

@end

File: assignments/java/anagram/src/test/java/AnagramTest.java
import static org.hamcrest.CoreMatchers.*;
import static org.junit.matchers.JUnitMatchers.*;
import static org.junit.Assert.*;

import java.util.*;
import org.junit.*;

public class AnagramTest {

    @Test
    public void testNoMatches() {
        Anagram detector = new Anagram("diaper");
        assertTrue(detector.match(Arrays.asList("hello", "world", "zombies", "pants")).isEmpty());
    }

    @Test
    public void testSimpleAnagram() {
        Anagram detector = new Anagram("ant");
        List<String> anagram = detector.match(Arrays.asList("tan", "stand", "at"));
        assertThat(anagram, hasItems("tan"));
    }

    @Test
    public void testDetectMultipleAnagrams() {
        Anagram detector = new Anagram("master");
        List<String> anagrams = detector.match(Arrays.asList("stream", "pigeon", "maters"));
        assertThat(anagrams, hasItems("maters", "stream"));
    }

    @Test
    public void testDoesNotConfuseDifferentDuplicates() {
        Anagram detector = new Anagram("galea");
        List<String> anagrams = detector.match(Arrays.asList("eagle"));
        assertTrue(anagrams.isEmpty());
    }

    @Test
    public void testIdenticalWordIsNotAnagram() {
        Anagram detector = new Anagram("corn");
        List<String> anagrams = detector.match(Arrays.asList("corn", "dark", "Corn", "rank", "CORN", "cron", "park"));
        assertThat(anagrams, hasItems("cron"));
    }

    @Test
    public void testEliminateAnagramsWithSameChecksum() {
        Anagram detector = new Anagram("mass");
        assertTrue(detector.match(Arrays.asList("last")).isEmpty());
    }

    @Test
    public void testEliminateAnagramSubsets() {
        Anagram detector = new Anagram("good");
        assertTrue(detector.match(Arrays.asList("dog", "goody")).isEmpty());
    }

    @Test
    public void testDetectAnagrams() {
        Anagram detector = new Anagram("listen");
        List<String> anagrams = detector.match(Arrays.asList("enlists", "google", "inlets", "banana"));
        assertThat(anagrams, hasItems("inlets"));
    }

    @Test
    public void testMultipleAnagrams() {
        Anagram detector = new Anagram("allergy");
        List<String> anagrams = detector.match(Arrays.asList("gallery", "ballerina", "regally", "clergy", "largely", "leading"));
        assertThat(anagrams, hasItems("gallery", "largely", "regally"));
    }

    @Test
    public void testAnagramsAreCaseInsensitive() {
        Anagram detector = new Anagram("Orchestra");
        List<String> anagrams = detector.match(Arrays.asList("cashregister", "Carthorse", "radishes"));
        assertThat(anagrams, hasItems("Carthorse"));
    }

}
